Add Mastermind project (#17)

3t8 opened the pull request:
Closes issue #17 

<!-- Congrats on completing your project tutorial draft! Before you submit your pull request, please check off the boxes on this checklist, like so: [x] -->

- [x] My project tutorial topic has already been aproved by the Codédex team
- [x] My project tutorial file renders properly in `mdx`
- [x] The content written is my own, and not plagiarized from other sources
- [x] All code written is commented or documented thoroughly
- [x] I have already completed a self-review of the project tutorial 
- [x] I will comply with changes from the Codédex team about my project tutorial

sonnynomnom commented on projects/create-mastermind-game-using-python/create-mastermind-game-using-python.mdx:
```diff
@@ -0,0 +1,225 @@
+  - python
+---
+
+# Create Mastermind game using Python
```
```suggestion
# Create Mastermind Game with Python
```

sonnynomnom commented on projects/create-mastermind-game-using-python/create-mastermind-game-using-python.mdx:
```diff
@@ -0,0 +1,225 @@
+---
+title: Create Mastermind game using Python
```
```suggestion
title: Create Mastermind Game with Python
```

sonnynomnom commented on projects/create-mastermind-game-using-python/create-mastermind-game-using-python.mdx:
```diff
@@ -0,0 +1,225 @@
+---
+title: Create Mastermind game using Python
+author: 3t8
```
Hi! Not sure if I mentioned this before, but is it possible to show your name here?

sonnynomnom commented on projects/create-mastermind-game-using-python/create-mastermind-game-using-python.mdx:
```diff
@@ -0,0 +1,225 @@
+- The number of digits that are correct and in the right position is returned
+- The number of digits that are correct but in the wrong position is returned
+
+## [#](#-generate-a-random-number) Step 1. Generate a random number
```
Let's make all the titles title cased.

```suggestion
## [#](#-generate-a-random-number) Step 1. Generate a Random Number
```

sonnynomnom commented on projects/create-mastermind-game-using-python/create-mastermind-game-using-python.mdx:
````diff
@@ -0,0 +1,225 @@
+print('You are a mastermind! You guessed the number in ' + str(tries) + ' tries')
+```
+
+## [#](#-improvements) improvements
````
```suggestion
## [#](#-improvements) Improvements
```

sonnynomnom commented on projects/create-mastermind-game-using-python/create-mastermind-game-using-python.mdx:
````diff
@@ -0,0 +1,225 @@
+
+```python
+from random import randrange
+# Generate a random number from 0 to 10000
````
```suggestion

# Generate a random number from 0 to 10000
```

Let's add some newlines

sonnynomnom reviewed (changes requested):
HOLY CRAP THIS IS A GOOD TUTORIAL. Love the setup. Super solid formating.

I added some small formatting things, feel free to make the changes and @Bobliuuu will take a look!

3t8 commented:
Thanks a lot!
I updated the project with your suggestions, found and fixed a few typos that I previously missed and ended up changing the name of the project according to the new title. Let me know if I should change anything else.
@sonnynomnom

sonnynomnom commented:
Alright, I just shared with the team this project, we all freaking love it. 

BUT we are wondering, do you think the black and white terminal version is better or a color version is better in your opinion?

Our engineer brought up that we could use color in the terminal with something like [colorama](https://pypi.org/project/colorama/).

But I just don't know if it's too late or not and if it's worth it. If you think this will make the project 2x better though, then I think we should do it!

3t8 commented:
Thanks a lot for the update. As I mentioned in the project, representing colors can be a bit tricky.

I'm guessing with something like colorama you could use ANSI color codes on windows too, but I'm not sure that would cover web-based python interpreters.

If you think a more classical version of the game with colors instead of numbers would be better I have no problem implementing that

Bobliuuu commented on projects/create-mastermind-game-with-python/create-mastermind-game-with-python.mdx:
```diff
@@ -0,0 +1,226 @@
+
+## [#](#-introduction) Introduction
+
+With more than 55 million units sold, [Mastermind](https://en.wikipedia.org/wiki/Mastermind_(board_game)) is a one of the world’s most popular strategy games ever. The idea behind it is to guess an unknown sequence of colored pegs in the fewest number of attempts possible. Despite having simple rules the game is still very difficult and entertaining.
```
```suggestion
With more than 55 million units sold, [Mastermind](https://en.wikipedia.org/wiki/Mastermind_(board_game)) is a one of the world’s most popular strategy games. The idea behind it is to guess an unknown sequence of colored pegs in the fewest number of attempts possible. Despite having simple rules, the game is still very difficult and entertaining.
```

Bobliuuu commented on projects/create-mastermind-game-with-python/create-mastermind-game-with-python.mdx:
```diff
@@ -0,0 +1,226 @@
+
+With more than 55 million units sold, [Mastermind](https://en.wikipedia.org/wiki/Mastermind_(board_game)) is a one of the world’s most popular strategy games ever. The idea behind it is to guess an unknown sequence of colored pegs in the fewest number of attempts possible. Despite having simple rules the game is still very difficult and entertaining.
+
+Because representing colors in a terminal is not an easy task, we are going to create a more *modern* version of the Mastermind game. In 1977 Invicta released an electronic version of the game where the colored pegs were replaced by an array of up to five digits. In this tutorial we are going to recreate it in Python.
```
```suggestion
Since representing colors in a terminal is not an easy task, we are going to create a more *modern* version of the Mastermind game. In 1977 Invicta released an electronic version of the game where the colored pegs were replaced by an array of up to five digits. In this tutorial, we are going to recreate that version in Python.
```

Bobliuuu commented on projects/create-mastermind-game-with-python/create-mastermind-game-with-python.mdx:
```diff
@@ -0,0 +1,226 @@
+
+The rules are really simple:
+- A random 4 digit secret code is generated
+- The player will input a 4 digit number trying to guess the secret code
```
```suggestion
- The player will input a 4 digit number to try to guess the secret code
```

Bobliuuu commented on projects/create-mastermind-game-with-python/create-mastermind-game-with-python.mdx:
```diff
@@ -0,0 +1,226 @@
+- A random 4 digit secret code is generated
+- The player will input a 4 digit number trying to guess the secret code
+- The number of digits that are correct and in the right position is returned
+- The number of digits that are correct but in the wrong position is returned
```
Nitpick, but could you add a small note here about what "right position" and "correct" means? 
The easiest way would be to use a simple mastermind example to walk through it.

Goku-kun replied:
+1

Bobliuuu commented on projects/create-mastermind-game-with-python/create-mastermind-game-with-python.mdx:
```diff
@@ -0,0 +1,226 @@
+
+## [#](#-read-users-guess) Step 2. Read user's guess
+
+The next step is getting user's guess. To do that we will use the `input` function. 
```
```suggestion
The next step is getting the user's guess. To do that, we will use the `input` function. 
```

Bobliuuu commented on projects/create-mastermind-game-with-python/create-mastermind-game-with-python.mdx:
```diff
@@ -0,0 +1,226 @@
+
+We use `isnumeric` to check that all the characters taken as input are numeric and `len(guess)` to make sure that the guess has 4 digits.
+
+If any of those conditions are not met, we will ask again the user for a guess.
```
```suggestion
If any of those conditions are not met, we will ask the user for a guess again.
```

Bobliuuu commented on projects/create-mastermind-game-with-python/create-mastermind-game-with-python.mdx:
````diff
@@ -0,0 +1,226 @@
+while (guess.isnumeric() == False or len(guess) != 4):
+  guess = input('Guess the four digit number: ')
+```
+The `guess` read with `input` is treated as a string, same as the secret code `number`.
````
Isn't `number` treated as a number here, not a string?

Goku-kun replied:
Yes. Please change `guess = 0` to `guess = '0'`.

Bobliuuu commented on projects/create-mastermind-game-with-python/create-mastermind-game-with-python.mdx:
````diff
@@ -0,0 +1,226 @@
+correct_position = 0
+```
+
+and one to count the correct digits in the wrong position
````
Move this to the previous line. 
"We need a variable ... in the right position and one ... wrong position.

Bobliuuu replied:
The two variables can also be combined to one python field.

Bobliuuu commented on projects/create-mastermind-game-with-python/create-mastermind-game-with-python.mdx:
```diff
@@ -0,0 +1,226 @@
+
+To improve this project you could try to:
+
+- Increase the number of digits the user has to guess. With 4 digits the possible permutations are 10000.
```
```suggestion
- Increase the number of digits the user has to guess. With 4 digits the possible permutations are 10000, but with more digits, it would be harder for the user to guess the permutation.
```

Bobliuuu reviewed (changes requested):
Love it! Just dropped a few small changes, please take a look 😄

asiqurrahman commented on projects/create-mastermind-game-with-python/create-mastermind-game-with-python.mdx:
```diff
@@ -0,0 +1,226 @@
+---
+title: Create Mastermind Game with Python
+author: 3t8
```
Are you ok with using your real name here?

If so please update your name here

Goku-kun commented on projects/create-mastermind-game-with-python/create-mastermind-game-with-python.mdx:
````diff
@@ -0,0 +1,226 @@
+```python
+correct_digit = 0
+```
+We will also use a copy of `number` in order to keep track of the digits that have already been matched and the ones that are left unmatched
````
Missing a period.

Goku-kun commented on projects/create-mastermind-game-with-python/create-mastermind-game-with-python.mdx:
```diff
@@ -0,0 +1,226 @@
+
+## [#](#-counting-the-number-of-guesses-and-winning) Step 7. Counting the number of guesses and winning
+
+The whole code after the `number` generation needs to be inside a
```
Need to finish this sentence with a colon or "`white` loop as follows:"

sonnynomnom commented:
Hi @3t8! Dharma just reviewed it one last time, can you take a look?

Also, after further thinking and research, let's do the colors!!! Just because colors is SUCH a big part of Mastermind

Let's use`colorama`, `termcolor`, or something else of your choice. Just because we should expect the user to use VS Code or Repl.it to finish this!

sonnynomnom commented:
Let me know when you finish these and we can merge it! @3t8

sonnynomnom commented:
take a look at this for inspiration: 
![mastermind](https://user-images.githubusercontent.com/4999022/206623204-ad622064-f15c-41f0-8e35-7f79c4e520ef.png)

sonnynomnom commented:
it doesn't have to be this crazy of course, but colors do add a layer

Bobliuuu commented:
Hey @3t8, how's progress on the changes? Please let us know if you need any help!
